Firestore/core/test/firebase/firestore/model/document_key_test.cc

```diff
@@ -0,0 +1,148 @@
+
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+TEST(DocumentKey, Constructor_Empty) {
+  const DocumentKey default_key;
+  EXPECT_TRUE(default_key.path().empty());
+
+  const auto& empty_key = DocumentKey::Empty();
+  const auto& another_empty_key = DocumentKey::Empty();
+  EXPECT_EQ(default_key, empty_key);
+  EXPECT_EQ(empty_key, another_empty_key);
+  EXPECT_EQ(&empty_key, &another_empty_key);
+}
+
+TEST(DocumentKey, Constructor_FromPath) {
+  ResourcePath path{"rooms", "firestore", "messages", "1"};
+  const DocumentKey key_from_path_copy{path};
+  // path shouldn't have been moved from.
+  EXPECT_FALSE(path.empty());
+  EXPECT_EQ(key_from_path_copy.path(), path);
+
+  const DocumentKey key_from_moved_path{std::move(path)};
+  EXPECT_TRUE(path.empty());
+  EXPECT_FALSE(key_from_moved_path.path().empty());
+  EXPECT_EQ(key_from_path_copy.path(), key_from_moved_path.path());
+}
+
+TEST(DocumentKey, CopyAndMove) {
+  DocumentKey key({"rooms", "firestore", "messages", "1"});
+  const std::string path_string = "rooms/firestore/messages/1";
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+
+  auto copied = key;
+  EXPECT_EQ(path_string, copied.path().CanonicalString());
+  EXPECT_EQ(key, copied);
+
+  const auto moved = std::move(key);
+  EXPECT_EQ(path_string, moved.path().CanonicalString());
+  EXPECT_NE(key, moved);
+  EXPECT_TRUE(key.path().empty());
+
+  // Reassignment.
+
+  key = copied;
+  EXPECT_EQ(copied, key);
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+
+  key = {};
+  EXPECT_TRUE(key.path().empty());
+  key = std::move(copied);
+  EXPECT_NE(copied, key);
+  EXPECT_TRUE(copied.path().empty());
+  EXPECT_EQ(path_string, key.path().CanonicalString());
+}
+
+TEST(DocumentKey, Constructor_StaticFactory) {
+  const auto key_from_segments =
+      DocumentKey::FromSegments({"rooms", "firestore", "messages", "1"});
+  const std::string path_string = "rooms/firestore/messages/1";
+  const auto key_from_string = DocumentKey::FromPathString(path_string);
+  EXPECT_EQ(path_string, key_from_string.path().CanonicalString());
+  EXPECT_EQ(path_string, key_from_segments.path().CanonicalString());
+  EXPECT_EQ(key_from_segments, key_from_string);
+}
+
+TEST(DocumentKey, Constructor_BadArguments) {
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey(ResourcePath{"foo"}), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey(ResourcePath{"foo", "bar", "baz"}), "");
+
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromSegments({"foo"}), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromSegments({"foo", "bar", "baz"}),
+                            "");
+
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString(""), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString("invalid//string"), "");
+  ASSERT_DEATH_IF_SUPPORTED(DocumentKey::FromPathString("invalid/key/path"),
+                            "");
+}
+
+TEST(DocumentKey, IsDocumentKey) {
+  EXPECT_TRUE(DocumentKey::IsDocumentKey({}));
```

Comment:
Curious. Do the other apis also consider an empty path to be a document key?

Reply:
They must, because they all implement the check as simple `path.size() % 2 == 0`. I'm not sure if they should, though, which is partially why I added this test. @wilhuff Gil, is this intentional?

Reply:
It's valid though hard to construct.

Document keys and collection keys differ only in even vs odd numbers of segments. Accordingly the zero segment key is a document key, though it's one you can neither read nor write.